File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.java
class Solution {
    public int helper(int i, int[] prices, int isBuying, int[][] visited) {
        if (i >= prices.length)
        return 0;

        if (visited[i][isBuying] != -1) return visited[i][isBuying];

        if (isBuying == 1) {
            int buy = helper(i + 1, prices, 0, visited) - prices[i];
            int cooldown = helper(i + 1, prices, 1, visited);
            visited[i][isBuying] = Math.max(buy, cooldown);
        } else {
            int sell = helper(i + 2, prices, 1, visited) + prices[i];
            int cooldown = helper(i + 1, prices, 0, visited);
            visited[i][isBuying] = Math.max(sell, cooldown);
        }

        return visited[i][isBuying];
    }

    public int maxProfit(int[] prices) {
        int[][] visited = new int[prices.length][2];
        
        for (int i = 0 ; i < prices.length; i++) {
            Arrays.fill(visited[i], -1);
        }

        return helper(0, prices, 1, visited);
    }
}

File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        memo = {}

        def dfs(i, choice):
            if (i == len(prices)):
                return 0

            if (i, choice) in memo:
                return memo[(i, choice)]
            '''
            Buy(0) -> profit -= prices[i]

            Sell(1) -> profit += prices[i]
                 -> profit = profit
            '''

            holdProfit = dfs(i + 1, choice)
            if (choice == 0): 
                # buy
                buyProfit = dfs(i + 1, 1) - prices[i]
                memo[(i, choice)] = max(buyProfit, holdProfit)
                return memo[(i, choice)]
            else:
                # sell
                sellProfit = dfs(i + 1, 0) + prices[i]
                memo[(i, choice)] = max(holdProfit, sellProfit)
                return memo[(i, choice)]
                
        return dfs(0, 0)